File: ch3/16_UVa10130.java
import java.util.*;

class Main { /* SuperSale */
  // 0-1 Knapsack DP (Top-Down) - faster as not all states are visited

  private static final int MAX_N = 1010;
  private static final int MAX_W = 40;
  private static int N, MW;
  private static int[] V = new int[MAX_N], W = new int[MAX_N];
  private static int[][] memo = new int[MAX_N][MAX_W];

  private static int value(int id, int w) {
    if (id == N || w == 0) return 0;
    if (memo[id][w] != -1) return memo[id][w];
    if (W[id] > w)         return memo[id][w] = value(id + 1, w);
    return memo[id][w] = Math.max(value(id + 1, w), V[id] + value(id + 1, w - W[id]));
  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int i, j, T, G, ans;

    T = sc.nextInt();
    while (T-- > 0) {
      for (i = 0; i < MAX_N; i++)
        for (j = 0; j < MAX_W; j++)
          memo[i][j] = -1;

      N = sc.nextInt();
      for (i = 0; i < N; i++) {
        V[i] = sc.nextInt();
        W[i] = sc.nextInt();
      }

      ans = 0;
      G = sc.nextInt();
      while (G-- > 0) {
        MW = sc.nextInt();
        ans += value(0, MW);
      }

      System.out.printf("%d\n", ans);
    }
  }
}

File: ch3/16_UVa10130.ml
(* SuperSale *)

(* 0-1 Knapsack DP (Top-Down) *)

open Printf
open Scanf

let max_n = 1010
let max_w = 40

let n = ref 0
let vs = Array.make max_n 0
let ws = Array.make max_n 0
let memo = Array.make_matrix max_n max_w (-1)

let rec value id w =
  if id = !n || w = 0 then 0
  else if memo.(id).(w) <> -1 then memo.(id).(w)
  else begin
    memo.(id).(w) <-
      if ws.(id) > w then value (id + 1) w
      else max (value (id + 1) w) (vs.(id) + value (id + 1) (w - ws.(id)));
    memo.(id).(w)
  end

let () =
  let tc = scanf "%d\n" (fun x -> x) in
  for _ = 1 to tc do
    Array.iter (fun a -> Array.fill a 0 (Array.length a) (-1)) memo;
    scanf "%d\n" (fun x -> n := x);
    for i = 0 to !n - 1 do
      scanf "%d %d\n" (fun x y -> vs.(i) <- x; ws.(i) <- y)
    done;
    let g = scanf "%d\n" (fun x -> x) in
    let ans = ref 0 in
    for _ = 1 to g do
      let mw = scanf "%d\n" (fun x -> x) in
      ans := !ans + value 0 mw
    done;
    printf "%d\n" !ans
  done



(* 0-1 Knapsack DP (Bottom-Up) *)

(* open Printf
 * open Scanf
 * 
 * let max_n = 1010
 * let max_w = 40
 * 
 * let () =
 *   let tc = scanf "%d\n" (fun x -> x) in
 *   for _ = 1 to tc do
 *     let n = scanf "%d\n" (fun x -> x) in
 *     let vs = Array.make (n + 1) 0 in
 *     let ws = Array.make (n + 1) 0 in
 *     for i = 1 to n do
 *       scanf "%d %d\n" (fun x y -> vs.(i) <- x; ws.(i) <- y)
 *     done;
 *     let c = Array.make_matrix max_n max_w 0 in
 *     for i = 1 to n do
 *       for w = 1 to max_w - 1 do
 *         c.(i).(w) <-
 *           if ws.(i) > w then c.(i - 1).(w)
 *           else max c.(i - 1).(w) (vs.(i) + c.(i - 1).(w - ws.(i)))
 *       done
 *     done;
 *     let g = scanf "%d\n" (fun x -> x) in
 *     let ans = ref 0 in
 *     for _ = 1 to g do
 *       let mw = scanf "%d\n" (fun x -> x) in
 *       ans := !ans + c.(n).(mw)
 *     done;
 *     printf "%d\n" !ans
 *   done *)
